### PR TITLE
Sparsity

### PR DESCRIPTION
**Pull request by rib-2**
Refactored @alexm-nm's work in the old private repo.

In addition to Alex's original work, I avoid materializing the dense matrices on CPU during weight loading (previous iteration materialized the entire model). Now, we create the dense matrices for loading on the fly.

Improvements:
- Currently, when we load the QKV matrix, we unpack and repack 3 times. We could make loading happen faster by unpacking during the first shard load and only repacking once all shards have been loaded
- ~~CUDAGraphs are not working~~ (fixed by https://github.com/neuralmagic/nm_gpu/pull/15)

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -144,6 +147,30 @@ def _verify_tokenizer_mode(self) -> None:
                 "either 'auto' or 'slow'.")
         self.tokenizer_mode = tokenizer_mode
 
+    def _verify_sparsity(self) -> None:
+        supported_sparsity = ["sparse_w16a16"]
```
Comment (alexm-nm) [inline vllm/config.py:151]: good idea to add w16a16!

File: vllm/config.py [hunk 2]
```diff
@@ -144,6 +147,30 @@ def _verify_tokenizer_mode(self) -> None:
...
+                raise ValueError(f"Unknown sparse method: {self.sparsity}. Must "
+                                 f"be one of {supported_sparse}.")
+        
+        hf_sparsity_config = getattr(self.hf_config, "sparsity_config", None)
+        if hf_sparsity_config is not None:
+            hf_sparsity_method = str(hf_sparse_config["sparse_method"]).lower()
+            if self.sparsity is None:
+                self.sparsity = hf_sparsity_method
+            elif self.sparsity != hf_sparsity_method:
+                raise ValueError(
```
Comment (alexm-nm) [inline vllm/config.py:168]: nice addition with validation

File: vllm/engine/arg_utils.py [hunk 3]
```diff
@@ -197,6 +198,15 @@ def add_cli_args(
...
                             'quantized and use `dtype` to determine the data '
                             'type of the weights.')
+        parser.add_argument('--sparsity',
+                            '-s',
+                            type=str,
+                            choices=['sparse_w16a16', None],
+                            default=None,
+                            help='Method used to compress sparse weights. If '
+                            'None, we first check the `sparsity_config` attribute '
+                            'in the model config gile. If that is None we assume '
```
Comment (alexm-nm) [inline vllm/engine/arg_utils.py:208]: gile => file

File: vllm/model_executor/layers/linear.py [hunk 4]
```diff
@@ -204,6 +205,10 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
         assert param_data.shape == loaded_weight.shape
         param_data.copy_(loaded_weight)
 
+        # If SparsParameter, repack dense data as sparse.
```
Comment (alexm-nm) [inline vllm/model_executor/layers/linear.py:208]: Spars => Sparse

File: vllm/model_executor/layers/linear.py [hunk 5]
```diff
@@ -260,9 +264,13 @@ def weight_loader(self,
                       param: Parameter,
                       loaded_weight: torch.Tensor,
                       loaded_shard_id: Optional[int] = None):
-        param_data = param.data
+        param_data = get_param_data(param)
         output_dim = getattr(param, "output_dim", None)
         if loaded_shard_id is None:
+            if isinstance(param, SparseParameter):
```
Comment (alexm-nm) [inline vllm/model_executor/layers/linear.py:270]: What's the reason for this limitation?

**Review by alexm-nm — APPROVED** [review]
LGTM! Thanks for the port

File: vllm/model_executor/layers/linear.py (hunk 5 above)
Comment (rib-2) [reply vllm/model_executor/layers/linear.py:270]: @alexm-nm - we should be able to support this. But I don't have / know of an example of a model yet which has this so I wasn't sure of how to test this

File: vllm/config.py (hunk 2 above)
Comment (rib-2) [reply vllm/config.py:168]: When we save the sparseml models, we are going to have a sparse_config in the hf config. once we have this, we can use this to load

File: vllm/model_executor/layers/linear.py (hunk 5 above)
Comment (alexm-nm) [reply vllm/model_executor/layers/linear.py:270]: Ok cool

**Review by alexm-nm — APPROVED** [review]

**Pull request merged**